src/prototype.rs

```diff
@@ -1766,25 +1765,22 @@ fn scan_uri(scanner: &mut InlineScanner) -> Option<String> {
 
 fn scan_email(scanner: &mut InlineScanner) -> Option<String> {
     // using a regex library would be convenient, but doing it by hand is not too bad
-
-    // FIXME: don't use strings?
     let mut uri: String = "mailto:".into();
 
     while let Some(c) = scanner.next() {
+        uri.push(c as char);
         match c {
-            c if is_ascii_alphanumeric(c) => uri.push(c as char),
+            c if is_ascii_alphanumeric(c) => (),
             b'.' | b'!' | b'#' | b'$' | b'%' | b'&' | b'\'' | b'*' | b'+' | b'/' |
-            b'=' | b'?' | b'^' | b'_' | b'`' | b'{' | b'|' | b'}' | b'~' | b'-' => uri.push(c as char),
-            b'@' => {
-                uri.push('@');
-                break;
-            }
-            _ => {
-                return None;
-            }
+            b'=' | b'?' | b'^' | b'_' | b'`' | b'{' | b'|' | b'}' | b'~' | b'-' => (),
+            _ => break,
         }
     }
 
+    if uri.as_bytes()[uri.len() - 1] != b'@' {
+        return None;
+    }
```

Comment:
And I think this changes the logic, leaves the `@` at the end. I didn't test it, but might be worth checking.

Reply:
Not sure of what you mean by this.. shouldn't `@` be at the end at that point?